realm/realm-library/src/objectServer/java/io/realm/PermissionManager.java

```diff
@@ -0,0 +1,525 @@
+                    }
+                }
+            });
+        }
+    }
+
+    private void checkIfRealmsAreOpenedAndRunDelayedTasks() {
+        synchronized (errorLock) {
+            if ((permissionRealm != null || permissionRealmError != null)
+                && (managementRealm != null || managementRealmError != null)) {
+                openInProgress = false;
+                runDelayedTasks();
+            }
+        }
+    }
+
+    private void checkCallbackNotNull(Callback<?> callback) {
+        if (callback == null) {
+            throw new IllegalArgumentException("Non-null 'callback' required.");
+        }
+    }
+
+    private boolean isReady() {
+        return managementRealm != null && permissionRealm != null;
+    }
+
+    private void checkIfValidThread() {
+        // Checks if we are in thread that created the PermissionManager.
+        if (threadId != Thread.currentThread().getId()) {
+            throw new IllegalStateException("PermissionManager was accessed from the wrong thread. It can only be " +
+                                            "accessed on the thread it was created on.");
+        }
+    }
+
+    /**
+     * FIXME Add Javadoc
+     */
+    @Override
+    public void close() {
+        checkIfValidThread();
+
+        // Multiple instances open, just decrement the reference count
+        Integer instanceCount = permissionManagerInstanceCounter.get();
+        if (instanceCount > 1) {
+            permissionManagerInstanceCounter.set(instanceCount - 1);
+            return;
+        }
+
+        // Only one instance open. Do a full close
+        permissionManagerInstanceCounter.set(0);
+        permissionManager.set(null);
+        delayedTasks.clear();
+
+        // If Realms are still being opened, abort that task
+        if (managementRealmOpenTask != null) {
+            managementRealmOpenTask.cancel();
+            managementRealmOpenTask = null;
+        }
+        if (permissionRealmOpenTask != null) {
+            permissionRealmOpenTask.cancel();
+            permissionRealmOpenTask = null;
+        }
+
+        // If Realms are opened. Close them.
+        if (managementRealm != null) {
+            managementRealm.close();
+        }
+        if (permissionRealm != null) {
+            permissionRealm.close();
+        }
+        closed = true;
+    }
+
+    /**
+     * Checks if this PermissionManager is closed or not. If it is closed, all methods will report back an error.
+     *
+     * @return {@code true} if the PermissionManager is closed, {@code false} if it is still open.
+     */
+    public boolean isClosed() {
+        checkIfValidThread();
+        return closed;
+    }
+
+    @Override
+    protected void finalize() throws Throwable {
+        if (managementRealm != null || permissionRealm != null) {
+            RealmLog.warn("PermissionManager was not correctly closed before being finalized.");
+            close();
```

Comment:
`finalize()` will run on the finalizer thread, calling `close()` will just crash because of the thread checking. Just remove the call to `close()`